src/sources/util/grpc/connectionlimit.rs

```diff
@@ -0,0 +1,110 @@
+use std::{
+    sync::{Arc, Mutex},
+    time::{Duration, Instant},
+    task::{Context, Poll},
+};
+
+use http::{Request, Response};
+use hyper::Body;
+use std::future::Future;
+use std::pin::Pin;
+
+use tonic::{
+    Status,
+    body::BoxBody,
+    transport::server::NamedService,
+};
+use tower::{Layer, Service};
+
+/// A service that tracks the number of requests and elapsed time,
+/// shutting down the connection gracefully if the configured limits are reached.
+#[derive(Clone)]
+pub struct ConnectionLimit<S> {
+    inner: S,
+    request_count: Arc<Mutex<usize>>,
+    max_requests: usize,
+    max_duration: Duration,
+    start_time: Instant,
+}
+
+impl<S> ConnectionLimit<S> {
+    pub fn new(inner: S, max_requests: usize, max_duration: Duration) -> Self {
+        Self {
+            inner,
+            request_count: Arc::new(Mutex::new(0)),
+            max_requests: max_requests,
+            max_duration: max_duration,
+            start_time: Instant::now(),
+        }
+    }
+}
+
+impl<S> Service<Request<Body>> for ConnectionLimit<S>
```

Comment:
I think this would end up applying one duration and request limit across all clients. In an ideal world, I think we would apply the limits per-client (that is per socket).